subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/DefaultScriptCompilationHandler.java

```diff
@@ -75,26 +78,35 @@
     private static final int HAS_METHODS_FLAG = 2;
 
     private final ClassLoaderCache classLoaderCache;
+    private final Deleter deleter;
     private final Map<String, List<String>> simpleNameToFQN;
 
-    public DefaultScriptCompilationHandler(ClassLoaderCache classLoaderCache, ImportsReader importsReader) {
+    public DefaultScriptCompilationHandler(ClassLoaderCache classLoaderCache, Deleter deleter, ImportsReader importsReader) {
         this.classLoaderCache = classLoaderCache;
-        simpleNameToFQN = importsReader.getSimpleNameToFullClassNamesMapping();
+        this.deleter = deleter;
+        this.simpleNameToFQN = importsReader.getSimpleNameToFullClassNamesMapping();
     }
 
     @Override
     public void compileToDir(ScriptSource source, ClassLoader classLoader, File classesDir, File metadataDir, CompileOperation<?> extractingTransformer,
                              Class<? extends Script> scriptBaseClass, Action<? super ClassNode> verifier) {
         Timer clock = Time.startTimer();
-        GFileUtils.deleteDirectory(classesDir);
-        GFileUtils.mkdirs(classesDir);
+        try {
+            deleter.ensureEmptyDirectory(classesDir, true);
+        } catch (IOException ioex) {
```

Comment:
Can we make the methods of `Deleter` not throw an IOException? Throwing an unchecked exception would make the usage easier.

Reply:
I looked at that and in several places it does make sense to catch the exception and add more context. Let's keep it this way for now.

Reply:
Nobody stops you from catching the exception to add more context even if we throw only an unchecked exception.